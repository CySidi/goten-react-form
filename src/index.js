import React, { Component } from 'react'

import './gotenForm.css'


const pubsubMessage = 'VALIDATE_GOTEN_COMPONENTS_'
const pubsubMessageResponse = '_RESPONSE'
let cont = 0

export class GotenForm extends Component {
    constructor(props) {
        super(props)
        this.pubsubMessage = pubsubMessage + ++cont
        this.gotenTextFieldCant = 0
        this.responses = 0
        this.errorsCant = 0
        this.subscription = PubSub.subscribe(this.pubsubMessage + pubsubMessageResponse, (_, response) => {
            if (response) {
                this.responses ++
            } else {
                return
            }
            if(this.responses === this.gotenTextFieldCant){
                this.props.onClick(event)
            }
        })
    }
    
    renderComponents(children) {
        return React.Children.map(children, child => {
            if (!React.isValidElement(child)) return child
            let childProps = {}
            if (child.type.name === 'GotenTextField') {
                this.gotenTextFieldCant ++
                childProps = { _pubsub_message: this.pubsubMessage }
                return React.cloneElement(child, childProps)
            }
            childProps.children = this.renderComponents(child.props.children)
            return React.cloneElement(child, childProps)
        })
    }


    onClick = (event) => {
        this.responses = 0
        PubSub.publish(this.pubsubMessage, null)
    }

    render() {
        this.gotenTextFieldCant = 0
        return (
            <div>
                {this.renderComponents(this.props.children)}
                {React.cloneElement(this.props.buttonComponent, {
                    onClick: this.onClick
                })}
            </div>
        )
    }

    componentWillUnmount() {
        PubSub.unsubscribe(this.subscribe)
    }
}