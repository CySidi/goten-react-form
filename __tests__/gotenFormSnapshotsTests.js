import React from 'react'
import renderer from 'react-test-renderer'

import { GotenTextField } from 'goten-react-text-field'
import { GotenForm } from '../src'


describe('GotenForm snapshots', () => {

    it('with onSuccess', () => {
        const tree = renderer.create(
            <GotenForm
                onSuccess={_ => null}
            />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
    
    it('with onError', () => {
        const tree = renderer.create(
            <GotenForm
                onSuccess={_ => null}
                onError={_ => null}
            />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
    
    it('with buttonComponent', () => {
        const tree = renderer.create(
            <GotenForm
                onSuccess={_ => null}
                buttonComponent={
                    <input
                        type='submit'
                        value='Boton'
                    />
                }
            />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
    
    it('with components', () => {
        const tree = renderer.create(
            <GotenForm
                onSuccess={_ => null}
            >
                <GotenTextField
                    value={'value'}
                    label={'label'}
                    type={'text'}
                />
                <GotenTextField
                    value={'value'}
                    label={'label'}
                    type={'text'}
                />
            </GotenForm>
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('with components and buttonComponent', () => {
        const tree = renderer.create(
            <GotenForm
                onSuccess={_ => null}
                buttonComponent={
                    <input
                        type='submit'
                        value='Boton'
                    />
                }
            >
                <GotenTextField
                    value={'value'}
                    label={'label'}
                    type={'text'}
                />
                <GotenTextField
                    value={'value'}
                    label={'label'}
                    type={'text'}
                />
            </GotenForm>
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
})