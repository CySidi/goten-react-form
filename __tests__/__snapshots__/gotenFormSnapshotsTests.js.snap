// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GotenForm snapshots with buttonComponent 1`] = `
<div>
  <input
    onClick={[Function]}
    type="submit"
    value="Boton"
  />
</div>
`;

exports[`GotenForm snapshots with components 1`] = `
<div>
  <div>
    <label
      className="label"
    >
      label 
    </label>
    <input
      className="undefined"
      label="label"
      onChange={[Function]}
      type="text"
      value="value"
    />
    <br />
  </div>
  <div>
    <label
      className="label"
    >
      label 
    </label>
    <input
      className="undefined"
      label="label"
      onChange={[Function]}
      type="text"
      value="value"
    />
    <br />
  </div>
</div>
`;

exports[`GotenForm snapshots with components and buttonComponent 1`] = `
<div>
  <div>
    <label
      className="label"
    >
      label 
    </label>
    <input
      className="undefined"
      label="label"
      onChange={[Function]}
      type="text"
      value="value"
    />
    <br />
  </div>
  <div>
    <label
      className="label"
    >
      label 
    </label>
    <input
      className="undefined"
      label="label"
      onChange={[Function]}
      type="text"
      value="value"
    />
    <br />
  </div>
  <input
    onClick={[Function]}
    type="submit"
    value="Boton"
  />
</div>
`;

exports[`GotenForm snapshots with onError 1`] = `<div />`;

exports[`GotenForm snapshots with onSuccess 1`] = `<div />`;
